### ISSUE TITLE
[SPIKE] Investigate TTL pros and cons

### ISSUE THREAD

**Issue by adelefelicia** [indykite/neo4j-graph-tool-core#175]
**Created by: @bayegan Mar 15 2022 at 15:46 CET**
Last updated: Jul 19 2022 at 11:08 CET
ClickUp ID: IK-2438

[SPIKE] Explore how to make further decision

This will only handle existing sessions
When new sessions are created, what should we do in that case?
Should be enabled at creation time?

Original idea: add the TTL whenever we enable TTL functions
Opposition: Performs better to run periodically 
Adding TTL in session creation - What are the pros and cons? Permission penalty?
Do we have experience using it or not? Input from  

In general, using apoc.ttl seems convenient to deal with nodes to expire at a given time. It is as simple as manually setting a label named :TTL and a property named ttl having the expiration timestamp; because of its simplicity, is also computationally very cheap in both initialisation and job management. The conveniency comes from the already built-in configurability regarding frequency of the job (in seconds) and limit of deleted nodes for each run.

In particular if used just for sessions at the moment (or to say just a few different nodes to check for), and if the configuration options are enough (I think they are), we wouldn't have to bother ourselves by setting up another in-db job scheduler (like apoc triggers), nor an off-db solution (I now think for example of an Airflow instance).
This is why I think it is not necessary to run an initialisation call periodically with a trigger, and it would make more sense to have it at the session's creation time. As already said, performance is not an issue here; rather, the strict configurability (basically cannot set any other label than :TTL , any other property than ttl ).

In conclusion, I think that at the moment, if no other use is found for periodic jobs other than for :Session nodes, apoc.ttl is more than enough. Also, not to imply other schedulers, the initialisation call of the procedure to set all the nodes to be :TTL  would have sense inside the session creation cypher.
[EDIT:Actually, we may also find ttl useful for regularly deleting automation test users that are older than n days. This would neither overload the db nor conflict with the ttl applied to :Session nodes. Simply, a mix of both could be deleted every day with a total limit of 20k nodes each run.]

:link: Linked tasks: #174,

**Comment by adelefelicia** [comment]
**Commented by: @indyboox Mar 23 2022 at 11:25 CET**

cc @ Bohumil Pospíšil could you please review this, also from the discussion we had? Thanks!
